### DISCUSSION TITLE
Line-in gain clipping

### DISCUSSION THREAD

**Discussion by nradina** [pschatzmann/arduino-audiokit#53] [Q&A] [answered]
Hello,

I've been experimenting with the "streams-audiokit-audiokit" while using different input level options. I'm encountering clipping of the input stage with anything over 100mV.

Curious if it's possible to adjust this parameter to a higher value (maybe the ES8388 has the ability to adjust the feedback of the input op amp?)

Thank you! Wonderful library!

**Comment by pschatzmann** [comment, marked as answer]
The ADF API does not provide any high level function, but you can try to play around with the functions on the driver level: 
e.g es8388_set_mic_gain(). I never used this, so you are on your own...

**Comment by nradina** [comment]
I was able to do some tests tonight regarding adjusting the gain on the AudioKit board.

Initially I noticed severe clipping, yet with help here from Phil as well as resources from jujax (https://github.com/jujax/lyrat_passthru) and thaarrak aka NA5Y (https://github.com/thaaraak/es8388/blob/master/src/es8388.cpp) I was able to get to a cleaner place.

It seems by default the PGA gain is set to +9 which was the cause of the clipping. 
I don't know if this is at the input to the ADC or the input to the amplifier chip.

There is a line in the es8338.c file that address this gain.
**299  res |= es_write_reg(ES8388_ADDR, ES8388_ADCCONTROL1, 0xbb); // MIC Left and Right channel PGA gain**

The path is: **arduino-audiokit/src/audio_driver/es8388/es8388.c**

0xbb sets the gain to 9db
0x00 sets the gain to 0.

6.2.1 Register 9 – ADC Control 1, Default 0000 0000
MicAmpL 7:4 Left channel PGA gain
0000 – 0 dB (default)
0001 – +3 dB
0010 – +6 dB
0011 – +9 dB
0100 – +12 dB
0101 – +15 dB
0110 – +18 dB
0111 – +21 dB
1000 – +24 dB

I was able to take a look on a scope of the clipping and max peak to peak while probing at the input pin of the speaker amplifier.
Playing around a bit with the kit.setVolume parameter I settled on this:
kit.setVolume(.96) which yields just about 940mV PP with a 1.20mV input signal.

(I added this line to the Kit-to-kit example after the kit.begin(cfg); line)

I also chose "0" for the volume hack parameter.


Attached are some photos that may do a better job of explanation.

I hope this is helpful to others. Thank you again, Phil!

Photos>>>>

![IMG_3298](https://user-images.githubusercontent.com/8020979/207761239-0d074b73-8025-4522-8378-74467df493ff.JPG)

![NewFile3](https://user-images.githubusercontent.com/8020979/207761311-a529f801-28e3-4b57-831f-d4d264146bb4.png)

![NewFile1](https://user-images.githubusercontent.com/8020979/207762170-a9c3613b-025d-4983-a87d-cb72ea8efd8d.png)

![Screen Shot 2022-12-14 at 9 25 13 PM](https://user-images.githubusercontent.com/8020979/207761330-397df506-b7a2-43aa-8adc-7291f4bfd478.png)

![Screen Shot 2022-12-14 at 9 25 22 PM](https://user-images.githubusercontent.com/8020979/207761331-5f9bcb68-9705-424c-8ae9-d0fb2c86dfea.png)

![Screen Shot 2022-12-14 at 9 25 47 PM](https://user-images.githubusercontent.com/8020979/207761332-cfbfab9f-bc83-4bf6-a50a-4d300e319ddc.png)

**Comment by pschatzmann** [comment]
You can easily figure out if the clipping is on the input side if you send the output to a CsvStream and look at the signal in the Arudino Serial Plotter: In fact as you can find in the documentation that is usually the first step to do!

ES8388_ADCCONTROL1 can be set to any required value by just calling ```es8388_set_mic_gain(es_mic_gain_t  gain) ``` which expects 
```
typedef enum {
    MIC_GAIN_MIN = -1,
    MIC_GAIN_0DB = 0,
    MIC_GAIN_3DB = 3,
    MIC_GAIN_6DB = 6,
    MIC_GAIN_9DB = 9,
    MIC_GAIN_12DB = 12,
    MIC_GAIN_15DB = 15,
    MIC_GAIN_18DB = 18,
    MIC_GAIN_21DB = 21,
    MIC_GAIN_24DB = 24,
    MIC_GAIN_MAX,
} es_mic_gain_t;
```
In addition I added the following to AudioKitSettings.h
```
// Define the default gain for the microphone amp 0xbb=9DB; 0x00=0DB
// Alternativly you can call es8388_set_mic_gain(es_mic_gain_t gain) if you prefer to use
// value from an comprehensive enum
#define ES8388_DEFAULT_MIC_GAIN         0xbb
```

**Reply by nradina** [reply]
Wonderful! Thank you, Phil!

**Comment by nradina** [comment]
I was able to update files today with the latest AudioKit code (Thank  you!) and test.
To start, I set the AudioKitSettings.h file to use 0x00 for input gain (0dB) and volume hack to 0.
Uploaded the streams-audiokit-serial sketch and launched the plotter to double check I wasn't clipping the ADC. 
All looked great on the plotter.

Although when probing the input pin of the amplifier, clipping is being shown. (1st image below)
I've played around with gain settings a bit in code (adjusting the volume hack, KitVolume, the new Default_Mic_Gain settings) but can't seem to find out why I see clipping at the analog input to the NS4150 amplifier).

When lowering my test signal, I was able to eliminate the clipping yet obviously the overall gain is reduced (2nd image below..adjusted the vertical scale).

A few questions:
1) With the "stream to serial" sketch, is this showing the ADC stream before the input to the DAC?
2) How does the volume hack work?
3) Am I totally missing an obvious principle with this excellent library?

Thanks for any help!
Nicholas

![NewFile8](https://user-images.githubusercontent.com/8020979/208151419-e6acbb0a-f7c8-4e82-8d7c-1cee3e263bed.png)
**1Vpp input signal, no clipping on serial plotter.**


![NewFile9](https://user-images.githubusercontent.com/8020979/208151422-aba11a58-ac3d-4c41-b7a9-e4b34c5cdde6.png)
**Lower input signal (note: changed vertical scale from image 1)**

**Comment by pschatzmann** [comment]
1. The stream to serial sketch is just directly printing the values received via I2S.
2. I never bothered to find out because I was not impacted. It was basically other users telling me that this fixed thier issue that the output was much too low on their board. The standard setting is based on the unmodified code from Espressif!
3. Since most of the I2S DACs and ADCs do not have any volume control I usually rely on my [software based solution](https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control)

ps. I never run my board at full volume because this is much too loud!
